pkg/app/piped/livestatestore/cloudrun/store.go

```diff
@@ -16,36 +16,82 @@ package cloudrun
 
 import (
 	"context"
+	"sync"
 
 	"go.uber.org/zap"
 
+	provider "github.com/pipe-cd/pipecd/pkg/app/piped/cloudprovider/cloudrun"
 	"github.com/pipe-cd/pipecd/pkg/config"
-	"github.com/pipe-cd/pipecd/pkg/model"
 )
 
-type applicationLister interface {
-	List() []*model.Application
+type store struct {
+	config        *config.CloudProviderCloudRunConfig
+	cloudProvider string
+	apps          map[string]provider.ServiceManifest
+	mu            sync.RWMutex
+	logger        *zap.Logger
+	client        provider.Client
 }
 
-type Store struct {
-	logger *zap.Logger
-}
-
-type Getter interface {
+func (s *store) addApp(sm provider.ServiceManifest) {
+	appID := sm.Labels()[provider.LabelApplication]
+	if appID == "" {
+		return
+	}
+	s.mu.Lock()
+	s.apps[appID] = sm
+	s.mu.Unlock()
 }
 
-func NewStore(cfg *config.CloudProviderCloudRunConfig, cloudProvider string, appLister applicationLister, logger *zap.Logger) *Store {
-	logger = logger.Named("cloudrun").
-		With(zap.String("cloud-provider", cloudProvider))
+func (s *store) run(ctx context.Context) error {
+	client, err := provider.DefaultRegistry().Client(ctx, s.cloudProvider, s.config, s.logger)
+	if err != nil {
+		s.logger.Error("failed to new cloudrun client: %v", zap.Error(err))
+		return err
+	}
 
-	return &Store{
-		logger: logger,
+	const maxLimit = 500
+	var cursor string
+	svc := make([]*provider.Service, 0, maxLimit)
+	for {
+		ops := &provider.ListOptions{
+			Limit:         maxLimit,
+			LabelSelector: provider.MakeManagedByPipedLabel(),
+			Cursor:        cursor,
+		}
+		// Cloud Run Admin API rate Limits
+		// https://cloud.google.com/run/quotas#api
+		v, next, err := client.List(ctx, ops)
+		if err != nil {
+			s.logger.Error("failed to list cloudrun services: %v", zap.Error(err))
+			return err
+		}
+		svc = append(svc, v...)
+		if next == "" {
+			break
+		}
+		cursor = next
+	}
+	for i := range svc {
+		sm, err := svc[i].ServiceManifest()
+		if err != nil {
+			s.logger.Error("failed to load cloudrun service into service manifest: %v", zap.Error(err))
+			continue
+		}
+		s.addApp(sm)
 	}
+	return nil
 }
 
-func (s *Store) Run(ctx context.Context) error {
-	s.logger.Info("start running cloudrun app state store")
+func (s *store) GetAppLiveServiceManifest(appID string) provider.ServiceManifest {
+	s.mu.RLock()
+	sv, ok := s.apps[appID]
+	s.mu.RUnlock()
```

Comment:
Btw, in this simple case, we can use atomic instead of the mutex.
You can take a look at this.
https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/apistore/applicationstore/store.go#L115

Reply:
Exactly, let me apply it.